mars/services/subtask/core.py

```diff
@@ -61,6 +61,15 @@ class Subtask(Serializable):
     priority: Tuple[int, int] = TupleField("priority", FieldTypes.int32)
     rerun_time: int = Int32Field("rerun_time")
     extra_config: dict = DictField("extra_config")
+    stage_id: str = StringField("stage_id")
+    # An unique and deterministic key for subtask compute logic. See logic_key in operator.py.
+    logic_id: str = StringField("logic_id")
+    # index for subtask with same compute logic.
+    index: int = Int32Field("index")
+    # parallelism for subtask with same compute logic.
+    parallelism: int = Int32Field("parallelism")
```

Comment:
Ths `logic_parallelism` won't change after a subtask graph is generated. So there won't be synchoronizing issue if I understand correctly?

Reply:
`parallelism` represents actual parallelism of subtasks with same logic, naming it as `logic_parallelism` seems a little wired for me? Do we need to change to `logic_parallelism`?

Reply:
> `parallelism` represents actual parallelism of subtasks with same logic, naming it as `logic_parallelism` seems a little wired for me? Do we need to change to `logic_parallelism`?

`parallelism` is actually a variable for subtask groups, not subtasks. Thus `parallelism` is not parallelism inside the subtask itself but a task-level definition. Thus clarification is needed.

> Ths `logic_parallelism` won't change after a subtask graph is generated. So there won't be synchoronizing issue if I understand correctly?

This assertion might be valid for now, but it may limit future implementations of dynamic splitting esp. when handling skewness in join operands, as you need to synchronize between different subtasks.

Reply:
data skew and dynamic splitting up is a complex topic. for example, in spark, `data skew` will be handled using following ways:
* Dynamically changes sort merge join into broadcast hash join.
* Dynamically coalesces partitions (combine small partitions into reasonably sized partitions) after shuffle exchange. Very small tasks have worse I/O throughput and tend to suffer more from scheduling overhead and task setup overhead. Combining small tasks saves resources and improves cluster throughput.
* Dynamically handles skew in sort merge join and shuffle hash join by splitting (and replicating if needed) skewed tasks into roughly evenly sized tasks.
* Dynamically detects and propagates empty relations.
This is a complex topic, I think we can keep current design, and deal with it in later PR when we do some dynamic re-scheduling?

Reply:
`parallelism` changed to `logic_parallelism `